### PR TITLE
Build Tags

### PR DESCRIPTION
**Pull request by zabawaba99**
I split up the the `_` imports into multiple files and added build tags to them so that we can exclude certain imports and have the binary build faster if we are only using one database driver.

Thoughts?

### CODE REVIEW

**Comment by Tonkpils** [comment]
Thoughts on pushing this upstream and getting their thoughts as well?

File: cmd/goose/pq.go [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// +build !godrv,!mysql,!sqlite
+
+package main
+
+// including pq
```
Comment (phemmer) [inline cmd/goose/pq.go:5]: such comment. much redundancy :-)

**Comment by phemmer** [comment]
I personally don't like the fact that specifying one driver disables all others. I think this violates the rule of least surprise.

**Comment by zabawaba99** [comment]
I disagree @phemmer . However, this doesn't give you the ability to specify more than one driver. By specifying two tags, I think you'd end up with none. What would you suggest as an alternative?

---

@Tonkpils , I have no problem pushing this code upstream. We'd just have to push this to a Bitbucket origin.

File: cmd/goose/pq.go (hunk 1 above)
Comment (zabawaba99) [inline cmd/goose/pq.go:5]: #golint

**Comment by phemmer** [comment]
Instead of default everything enabled, then you specify one tag, and all but that one driver gets disabled, how about flags to disable the specific driver? For example `// +build !no-sqlite`

**Comment by phemmer** [comment]
We wouldn't be able to push this upstream as upstream imports the drivers within the library. Our code imports the drivers within the cli component.
Also upstream doesn't seem keen on merging PRs (which is why we have our own fork).

**Comment by phemmer** [comment]
As another requirement for this type of change, I think we need a way for the binary to indicate what drivers it was compiled with.

**Comment by zabawaba99** [comment]
> Instead of default everything enabled, then you specify one tag, and all but that one driver gets disabled, how about flags to disable the specific driver? For example // +build !no-sqlite>

I would be ok with that, so long as I can exclude them

> As another requirement for this type of change, I think we need a way for the binary to indicate what drivers it was compiled with.

The only way I can see of doing that would be to install the binary with specific build flags (`go install -ldflags "-X goose.libraries pq,mysql" .`).

**Comment by phemmer** [comment]
> The only way I can see of doing that would be to install the binary with specific build flags

You're creating driver-specific `.go` files, just add an `init()`

`go-sqlite3.go`:

```
// +build !godrv,!pq,!mysql

package main

// including go-sqlite3
import _ "github.com/mattn/go-sqlite3"

func init() {
 drivers = append(drivers, "go-sqlite3")
}
```

...where `drivers` is a global.

**Comment by zabawaba99** [comment]
``` bash
$ go install -tags "nopq" -a ./cmd/goose
$ goose drivers
Drivers:
    sqlite
    godrv
    mysql
$ go install -tags "nosqlite" -a ./cmd/goose
$ goose drivers
Drivers:
    godrv
    mysql
    pq
```

File: cmd/goose/godrv.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// +build !nogodrv
+
+package main
+
+// including godrv
+import _ "github.com/ziutek/mymysql/godrv"
+
+func init() {
+	drivers = append(drivers, "godrv")
```
Comment (phemmer) [inline cmd/goose/godrv.go]: Lets call this `MyMySQL` as that's what the driver calls itself.

Edit: or `mymysql`. That might make more sense so the tag can be lowercase and it'll match, but either is fine by me.

File: cmd/goose/go_sqlite3.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// +build !nosqlite
+
+package main
+
+// including go-sqlite3
+import _ "github.com/mattn/go-sqlite3"
+
+func init() {
+	drivers = append(drivers, "sqlite")
```
Comment (phemmer) [inline cmd/goose/go_sqlite3.go]: This should be called `sqlite3` (there is actually a difference :-P)

File: cmd/goose/go_sqlite3.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+// +build !nosqlite
```
Comment (phemmer) [inline cmd/goose/go_sqlite3.go]: The tag should match the driver name; `sqlite3`

File: cmd/goose/godrv.go [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// +build !nogodrv
```
Comment (phemmer) [inline cmd/goose/godrv.go]: ditto

Comment (zabawaba99) [inline cmd/goose/godrv.go]: amended and rebased

File: cmd/goose/go_sqlite3.go (hunk 4 above)
Comment (zabawaba99) [inline cmd/goose/go_sqlite3.go]: amended and rebased

**Comment by phemmer** [comment]
LGTM

File: cmd/goose/go_sqlite3.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// +build !nosqlite3
+
+package main
+
+// including go-sqlite3
```
Comment (Tonkpils) [inline cmd/goose/go_sqlite3.go:5]: I'd say this comment is superfluous and ditto for the rest of the drivers.

Comment (phemmer) [inline cmd/goose/go_sqlite3.go:5]: I commented on the same thing earlier (not sure where it went). Apparently it's a go lint thing.

Comment (zabawaba99) [inline cmd/goose/go_sqlite3.go:5]: Aye. It's why I left it in, to keep `go lint` nice and quiet. Although, we don't _need_ to always abide by what it says.

**Pull request merged**